R/download_file_linter.R

```diff
@@ -0,0 +1,83 @@
+#' Recommend usage of a portable `mode` value in [download.file()]
+#'
+#' `mode = "w"` or `mode = "a"` in `download.file()` can generate broken files
+#' on Windows. Instead, `?download.file()` recommends the usage of `mode = "wb"`
+#' and `mode = "ab"`.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "download.file(x = my_url)",
+#'   linters = download_file_linter()
+#' )
+#'
+#' lint(
+#'   text = "download.file(x = my_url, mode = 'w')",
+#'   linters = download_file_linter()
+#' )
+#'
+#' lint(
+#'   text = "download.file(x = my_url, , , , 'w')",
+#'   linters = download_file_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "download.file(x = my_url, mode = 'wb')",
+#'   linters = download_file_linter()
+#' )
+#'
+#' lint(
+#'   text = "download.file(x = my_url, , , , 'wb')",
+#'   linters = download_file_linter()
+#' )
+#'
+#' @evalRd rd_tags("download_file_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+download_file_linter <- function() {
+
+  Linter(linter_level = "expression", function(source_expression) {
+    xml_calls <- source_expression$xml_find_function_calls("download.file")
+
+    # 3 options:
+    # - no explicit value for mode
+    # - mode = "a" or mode = "w"
+    # - similar as previous but specified by position
+    xml_calls_no_mode <- xml_find_all(
+      xml_calls,
+      "parent::expr[1][
+        not(SYMBOL_SUB[text() = 'mode'])
+        and count(OP-COMMA) < 4
+      ]"
+    )
+
+    xml_calls_bad_mode_by_name <- xml_find_all(
+      xml_calls,
+      glue(
+        "parent::expr[1][
+          SYMBOL_SUB[text() = 'mode'
+          and following-sibling::expr[1]/STR_CONST[not(contains(text(), 'b'))]]
+        ]"
+      )
+    )
+
+    xml_calls_bad_mode_by_position <- xml_find_all(
```

Comment:
Unfortunately positional matching is not so easy, c.f.

```r
download.file(cacheOK=TRUE, destfile, method, quiet, x='http://...')
```

Will throw a lint in the current logic based on whether the URL has a 'b' in it.

Given how messy positional matching is, unles we have good evidence of people actually using the argument positionally, I would leave this for future work for now.